### PR TITLE
[PAL/Linux-SGX] Fix sysfs socket sanitization check

### PR DESCRIPTION
**Pull request by vijaydhanraj**
## Description of the changes 
In graphene, the current assumption is that number of sockets is equal to the number of nodes which is true for most Intel platforms. But virtual environments like QEMU can have each core in separate socket/package and just have one numa node. For such cases check against number of online logical processors instead of number of numa nodes.

PS: Although checking against numa nodes against socket was the intention, current code checks against num of cache index which is incorrect.

Signed-off-by: Vijay Dhanraj <vijay.dhanraj@intel.com>

The figure below shows the case under virtual environment like QEMU.
VM_Centos_topo

vs Baremetal Server
baremetal_Ubuntu_topo

## How to test this PR? 
Please run LibOS regression tests under virtual environments like QEMU.

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)

### CODE REVIEW

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @vijaydhanraj)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 342 at r1](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVJXfIH5GNKWxLXvMXa:-MVJXfIH5GNKWxLXvMXb:b-e8o9z9) ([raw file](https://github.com/oscarlab/graphene/blob/6606944bbef4e7daaf20a29f597def7352701661/Pal/src/host/Linux-SGX/db_main.c#L342)):*
> ```C
>     for (int64_t idx = 0; idx < num_cores; idx++) {
>         /* Virtual environements such as QEMU can have each core in separate socket/package and just
>          * have one numa node. For such cases check against number of online logical processors */
> ```

Is the number of sockets specified anywhere in sysfs? If so, it would be much better to check against this value than the number of cores.

**Review by vijaydhanraj — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 342 at r1](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVJXfIH5GNKWxLXvMXa:-MVJdKgN1qBSaa2tYiEu:bj3dhsg) ([raw file](https://github.com/oscarlab/graphene/blob/6606944bbef4e7daaf20a29f597def7352701661/Pal/src/host/Linux-SGX/db_main.c#L342)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Is the number of sockets specified anywhere in sysfs? If so, it would be much better to check against this value than the number of cores.
</blockquote></details>

I did look at this https://elixir.bootlin.com/linux/v4.15.18/source/Documentation/ABI/testing and couldn't find anything that provides the number of sockets.  We can start storing unique values as we read `/sys/devices/system/cpu/cpu0/topology/physical_package_id` to get the count.  But since we don't trust the host, not sure if this will be helpful.

**Review by mkow — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 342 at r1](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVJXfIH5GNKWxLXvMXa:-MVLjDF_CpS3S6pK47iq:b-5yjjtp) ([raw file](https://github.com/oscarlab/graphene/blob/6606944bbef4e7daaf20a29f597def7352701661/Pal/src/host/Linux-SGX/db_main.c#L342)):*
<details><summary><i>Previously, vijaydhanraj (Vijay Dhanraj) wrote…</i></summary><blockquote>

I did look at this https://elixir.bootlin.com/linux/v4.15.18/source/Documentation/ABI/testing and couldn't find anything that provides the number of sockets.  We can start storing unique values as we read `/sys/devices/system/cpu/cpu0/topology/physical_package_id` to get the count.  But since we don't trust the host, not sure if this will be helpful.
</blockquote></details>

Thanks for checking, resolving then.

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @vijaydhanraj)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 342 at r1](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVLyZl66fZEkM_61PhZ:-MVLyZl66fZEkM_61Ph_:b2l8usr) ([raw file](https://github.com/oscarlab/graphene/blob/6606944bbef4e7daaf20a29f597def7352701661/Pal/src/host/Linux-SGX/db_main.c#L342)):*
> ```C
>     for (int64_t idx = 0; idx < num_cores; idx++) {
>         /* Virtual environements such as QEMU can have each core in separate socket/package and just
>          * have one numa node. For such cases check against number of online logical processors */
> ```

You have a typo in `environments`. But also, I would rephrase to this:
```
        /* Virtual environments such as QEMU may assign each core to a separate socket/package.
         * To support such unusual topologies, we check against the number of online logical cores. */
```

**Review by vijaydhanraj — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: 0 of 1 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @mkow)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 342 at r1](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVLyZl66fZEkM_61PhZ:-MVN61dOEYSNQLrpDgCU:bb6un1j) ([raw file](https://github.com/oscarlab/graphene/blob/6606944bbef4e7daaf20a29f597def7352701661/Pal/src/host/Linux-SGX/db_main.c#L342)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

You have a typo in `environments`. But also, I would rephrase to this:
```
        /* Virtual environments such as QEMU may assign each core to a separate socket/package.
         * To support such unusual topologies, we check against the number of online logical cores. */
```
</blockquote></details>

Done

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @vijaydhanraj)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 342 at r2](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVNEs2v0ut3ykSSYU1u:-MVNEs2v0ut3ykSSYU1v:bw7p5pt) ([raw file](https://github.com/oscarlab/graphene/blob/0324ccef8a11f24c63d0b0f22bca7b763a58b8a6/Pal/src/host/Linux-SGX/db_main.c#L342)):*
> ```C
>     for (int64_t idx = 0; idx < num_cores; idx++) {
>         /* Virtual environments such as QEMU may assign each core to a separate socket/package with
>          * one or more numa node. To support such unusual topologies, we check against the number
> ```

`numa node.` -> `NUMA nodes.`
___
*[Pal/src/host/Linux-SGX/db_main.c, line 343 at r2](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVNEq-fDA_-6uulxCVH:-MVNEq-fDA_-6uulxCVI:b-6w324r) ([raw file](https://github.com/oscarlab/graphene/blob/0324ccef8a11f24c63d0b0f22bca7b763a58b8a6/Pal/src/host/Linux-SGX/db_main.c#L343)):*
> ```C
>         /* Virtual environments such as QEMU may assign each core to a separate socket/package with
>          * one or more numa node. To support such unusual topologies, we check against the number
>          * of online logical cores */
> ```

Add a dot at the end.

**Review by vijaydhanraj — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: 0 of 1 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 342 at r2](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVNEs2v0ut3ykSSYU1u:-MVNILBs3pS3ZgnmvPNc:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/0324ccef8a11f24c63d0b0f22bca7b763a58b8a6/Pal/src/host/Linux-SGX/db_main.c#L342)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`numa node.` -> `NUMA nodes.`
</blockquote></details>

Done.
___
*[Pal/src/host/Linux-SGX/db_main.c, line 343 at r2](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVNEq-fDA_-6uulxCVH:-MVNIMVS1eFvRzORmrJY:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/0324ccef8a11f24c63d0b0f22bca7b763a58b8a6/Pal/src/host/Linux-SGX/db_main.c#L343)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Add a dot at the end.
</blockquote></details>

Done.

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @vijaydhanraj)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 343 at r3](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVOsVcp59_-sthZAIaO:-MVOsVcp59_-sthZAIaP:buxk5e2) ([raw file](https://github.com/oscarlab/graphene/blob/ba50b428a8d4e0aa3f7db19c62d80ecf84b9a2a7/Pal/src/host/Linux-SGX/db_main.c#L343)):*
> To support such unusual topologies, we check against the number of online logical cores.

This sentence doesn't make too much sense logically - now it says that we do this check to support this case. But we do this check for security, if we deleted this check altogether this case would be still supported :)
What you want to say here is that we only check against `num_cores` instead of `num_nodes` because we have to support this case.

**Review by vijaydhanraj — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @mkow)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 343 at r3](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVOsVcp59_-sthZAIaO:-MVPAyqg8tll7hEsdxRP:b-yfnr0j) ([raw file](https://github.com/oscarlab/graphene/blob/ba50b428a8d4e0aa3f7db19c62d80ecf84b9a2a7/Pal/src/host/Linux-SGX/db_main.c#L343)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

> To support such unusual topologies, we check against the number of online logical cores.

This sentence doesn't make too much sense logically - now it says that we do this check to support this case. But we do this check for security, if we deleted this check altogether this case would be still supported :)
What you want to say here is that we only check against `num_cores` instead of `num_nodes` because we have to support this case.
</blockquote></details>

Technically it will be incorrect to check number of sockets against number of numa nodes as they need not be same. But this statement `check against num_cores instead of num_nodes`  seems to imply that we should check against `num_nodes`

All I am trying to convey here is that why we are sanitizing against `num_cores`. 

How about this?
```
/* Virtual environments such as QEMU may assign each core to a separate socket/package with
 * one or more NUMA nodes. So we check against the number of online logical cores. */
```

**Review by dimakuv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @mkow)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 343 at r3](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVOsVcp59_-sthZAIaO:-MVQ1X1U4O3OH_1RimJE:b5fcfmx) ([raw file](https://github.com/oscarlab/graphene/blob/ba50b428a8d4e0aa3f7db19c62d80ecf84b9a2a7/Pal/src/host/Linux-SGX/db_main.c#L343)):*
<details><summary><i>Previously, vijaydhanraj (Vijay Dhanraj) wrote…</i></summary><blockquote>

Technically it will be incorrect to check number of sockets against number of numa nodes as they need not be same. But this statement `check against num_cores instead of num_nodes`  seems to imply that we should check against `num_nodes`

All I am trying to convey here is that why we are sanitizing against `num_cores`. 

How about this?
```
/* Virtual environments such as QEMU may assign each core to a separate socket/package with
 * one or more NUMA nodes. So we check against the number of online logical cores. */
```
</blockquote></details>

I'm fine with the suggested version of the comment.

**Review by mkow — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @vijaydhanraj)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 343 at r3](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVOsVcp59_-sthZAIaO:-MVSCPO86MvBsiagcK1i:bpiw74a) ([raw file](https://github.com/oscarlab/graphene/blob/ba50b428a8d4e0aa3f7db19c62d80ecf84b9a2a7/Pal/src/host/Linux-SGX/db_main.c#L343)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I'm fine with the suggested version of the comment.
</blockquote></details>

the proposed version sounds ok

**Review by vijaydhanraj — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @mkow)

___
*[Pal/src/host/Linux-SGX/db_main.c, line 343 at r3](https://reviewable.io/reviews/oscarlab/graphene/2218#-MVOsVcp59_-sthZAIaO:-MVSPpNy8Nkf3qkzR3Sv:b-16kztj) ([raw file](https://github.com/oscarlab/graphene/blob/ba50b428a8d4e0aa3f7db19c62d80ecf84b9a2a7/Pal/src/host/Linux-SGX/db_main.c#L343)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

the proposed version sounds ok
</blockquote></details>

Done, I have updated the comment.

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners

**Review by dimakuv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners

**Comment by dimakuv** [comment]
Jenkins, test this please

**Comment by vijaydhanraj** [comment]
Please let me know if I need to rebase to the latest master.

**Review by dimakuv — APPROVED** [review]
No need, we'll do it ourselves. We are waiting until 5 days elapse, otherwise we need a third review :)

> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218#-:-MVaksAI16gcLbKAHIht:b-v56wix)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners

**Comment by mkow** [comment]
Jenkins, retest this please

**Review by mkow — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL)

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2218)* status: :shipit: complete! all files reviewed, all discussions resolved
